refactor(plugin-server): Run ingestion only on worker threads

macobo opened the pull request:
## Problem

WIP. Creating this to run tests

Our ingestion logic is hard to understand and harder to split up. Doing some work in main thread and some in workers is one part of the problem.

## Changes

We now run all of the previous `ingestEvent` in the worker thread.

No behavior is dropped - most code is copy-pasta with minor callsite changes.

Once this is live we will have info whether doing ingestion purely in workers causes issues.

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

See tests.

macobo committed:
Refactor method to do everything it's supposed to do

No tests though?

macobo committed:
Remove unused argument from a function

macobo committed:
Remove dead code

macobo committed:
Run event pipeline on worker threads

macobo committed:
Remove code in way of refactor

yakkomajuri commented on plugin-server/src/worker/ingestion/event-pipeline.ts:
```diff
@@ -0,0 +1,84 @@
+                event: processedEvent,
+                func: async (event) => {
+                    const result = await ingestEvent(server, event)
+                    actionMatches = result.actionMatches || []
```
maybe for a follow up refactor but will just leave the comment here so I don't forget - would be great to separate out the action handling code into its own function rather than `ingestEvent` if we're doing this pipeline approach

macobo replied:
See comment above.

yakkomajuri commented on plugin-server/src/worker/ingestion/event-pipeline.ts:
```diff
@@ -0,0 +1,84 @@
+            }
+            await Promise.all(promises)
+        }
+    } else {
```
given this function has shrunk a bit might make sense to flip this on its head and have:

```js
if (!processedEvent) {
        // processEvent might not return an event. This is expected and plugins, e.g. downsample plugin uses it.
        server.statsd?.increment('kafka_queue.dropped_event', {
            teamID: String(event.team_id),
        })
        return
}
```
and then all the rest outside of the conditional.

Happy to keep as is though if it helps us see that we've just moved code and not changed it

macobo replied:
See comment above.

yakkomajuri reviewed (commented):
nice work - let's ofc get some tests in as well.

left some initial comments

macobo commented on plugin-server/src/worker/ingestion/event-pipeline.ts:
```diff
@@ -0,0 +1,84 @@
+
+        server.statsd?.increment('kafka_queue_single_event_processed_and_ingested')
+
+        if (actionMatches.length > 0) {
```
I'm planning on rewriting all of this code but separately. The goal with this PR is to answer the question of "will it work" easy and without doubts around if some logic broke. Cutting code from A -> B is the best way around that.

macobo commented on plugin-server/src/worker/ingestion/ingest-event.ts:
```diff
@@ -98,7 +96,16 @@ async function handleActionMatches(
+    person: Person | undefined,
+    teamId: TeamId
+): boolean {
+    if (!hub.CONVERSION_BUFFER_ENABLED && !hub.conversionBufferEnabledTeams.has(teamId)) {
```
Fair. Will move this after the statsd block.

macobo commented on plugin-server/src/worker/tasks.ts:
```diff
@@ -39,6 +40,9 @@ export const workerTasks: Record<string, TaskRunner> = {
     getPluginSchedule: (hub) => {
         return hub.pluginSchedule
     },
+    runEventPipeline: async (hub, args: { event: PluginEvent }) => {
```
See comment above. Not in scope for this PR.

macobo commented:
> nice work - let's ofc get some tests in as well.

Not sure what do you mean - what tests do we need here beyond what we already have?

Note the intent here (as discussed on tuesday) is 0 behavioral changes besides moving some code from outside worker to inside so we know it's safe (can run it in production). Actual refactoring and behavioral changes will happen once we have verified that it works and are happy with it.

macobo committed:
Make metrics happy

macobo committed:
Silence a yeetcode-related failure

macobo committed:
Mock over stopping unneeded services

macobo committed:
Yeetcode a test

macobo committed:
Reset the right table between tests

macobo committed:
Reset healthcheck consumer between tests

macobo committed:
Silence logspam around creating topics

macobo committed:
Significantly speed up resetting clickhouse tables between tests

This takes the function from 13-17s to 4s locally. Not great but not
terrible either.

macobo commented:
This is now ready for review/merge. @yakkomajuri please take a look and let's merge it sometime today :)

I needed to unfortunately include some test refactorings - there were some bugs in the test suite causing pain getting this green.

yakkomajuri reviewed (approved):
let's keep a close eye on this and revert if necessary.

macobo commented on plugin-server/tests/postgres/worker.test.ts:
```diff
@@ -163,7 +163,7 @@ describe('worker', () => {
         })
 
-        it('pauses the queue if too many tasks', async () => {
+        it.skip('pauses the queue if too many tasks', async () => {
```
Fair. I wasn't sure whether this was actually "dead" or not for clickhouse code, but looking at it again it is.

macobo committed:
Remove try-catch

macobo committed:
Remove a dead test
